deeplearning4j-scaleout/spark/dl4j-spark-parameterserver/src/main/java/org/deeplearning4j/spark/parameterserver/training/SharedTrainingMaster.java

```diff
@@ -0,0 +1,1044 @@
+import org.apache.spark.broadcast.Broadcast;
+import org.apache.spark.input.PortableDataStream;
+import org.apache.spark.storage.StorageLevel;
+import org.deeplearning4j.api.storage.StatsStorageRouter;
+import org.deeplearning4j.berkeley.Pair;
+import org.deeplearning4j.exception.DL4JInvalidConfigException;
+import org.deeplearning4j.optimize.api.IterationListener;
+import org.deeplearning4j.spark.api.*;
+import org.deeplearning4j.spark.api.stats.SparkTrainingStats;
+import org.deeplearning4j.spark.api.worker.ExecuteWorkerFlatMap;
+import org.deeplearning4j.spark.api.worker.ExecuteWorkerPDSFlatMap;
+import org.deeplearning4j.spark.api.worker.ExecuteWorkerPathFlatMap;
+import org.deeplearning4j.spark.api.worker.NetBroadcastTuple;
+import org.deeplearning4j.spark.impl.graph.SparkComputationGraph;
+import org.deeplearning4j.spark.impl.multilayer.SparkDl4jMultiLayer;
+import org.deeplearning4j.spark.impl.paramavg.BaseTrainingMaster;
+import org.deeplearning4j.spark.impl.paramavg.ParameterAveragingTrainingMaster;
+import org.deeplearning4j.spark.impl.paramavg.ParameterAveragingTrainingResult;
+import org.deeplearning4j.spark.impl.paramavg.stats.ParameterAveragingTrainingMasterStats;
+import org.deeplearning4j.spark.parameterserver.conf.SharedTrainingConfiguration;
+import org.deeplearning4j.spark.parameterserver.functions.*;
+import org.deeplearning4j.spark.parameterserver.networking.SilentTrainingDriver;
+import org.deeplearning4j.spark.util.SparkUtils;
+import org.nd4j.linalg.dataset.DataSet;
+import org.nd4j.linalg.dataset.api.MultiDataSet;
+import org.nd4j.parameterserver.distributed.VoidParameterServer;
+import org.nd4j.parameterserver.distributed.conf.VoidConfiguration;
+import org.nd4j.parameterserver.distributed.enums.ExecutionMode;
+import org.nd4j.parameterserver.distributed.enums.NodeRole;
+import org.nd4j.parameterserver.distributed.enums.TransportType;
+import org.nd4j.parameterserver.distributed.transport.MulticastTransport;
+import org.nd4j.parameterserver.distributed.transport.RoutedTransport;
+import org.nd4j.parameterserver.distributed.transport.Transport;
+import org.nd4j.shade.jackson.annotation.JsonAutoDetect;
+import org.nd4j.shade.jackson.annotation.PropertyAccessor;
+import org.nd4j.shade.jackson.core.JsonFactory;
+import org.nd4j.shade.jackson.core.JsonProcessingException;
+import org.nd4j.shade.jackson.databind.DeserializationFeature;
+import org.nd4j.shade.jackson.databind.MapperFeature;
+import org.nd4j.shade.jackson.databind.ObjectMapper;
+import org.nd4j.shade.jackson.databind.SerializationFeature;
+import org.nd4j.shade.jackson.dataformat.yaml.YAMLFactory;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * @author raver119@gmail.com
+ */
+@Slf4j
+public class SharedTrainingMaster extends BaseTrainingMaster<SharedTrainingResult, SharedTrainingWorker> implements TrainingMaster<SharedTrainingResult, SharedTrainingWorker> {
```

Comment:
And yes, as i see - we want as close data amount per executor (not worker!) as possible.